client/material/text-field.tsx

```diff
@@ -0,0 +1,373 @@
+  ) => {
+    const id = useId()
+    const [isFocused, setIsFocused] = useState(false)
+    const [inputRef, setInputRef] = useMultiRef<HTMLInputElement>(ref)
+
+    const autoSize = useCallback(
+      (elem: HTMLInputElement) => {
+        const padding = dense ? TEXTAREA_BOTTOM_PADDING_DENSE : TEXTAREA_BOTTOM_PADDING
+        // Needed in order to lower the height when deleting text
+        elem.style.height = `${rows * 20 + padding}px`
+        elem.style.height = `${elem.scrollHeight}px`
+        // Textarea doesn't scroll completely to the end when adding a new line, just to the
+        // baseline of the added text it seems, so we scroll manually to the end here
+        elem.scrollTop = elem.scrollHeight
+      },
+      [dense, rows],
+    )
+
+    const onInputBlur = useCallback(
+      (event: React.FocusEvent<HTMLInputElement>) => {
+        setIsFocused(false)
+        onBlur?.(event)
+      },
+      [onBlur],
+    )
+
+    const onInputFocus = useCallback(
+      (event: React.FocusEvent<HTMLInputElement>) => {
+        setIsFocused(true)
+        onFocus?.(event)
+      },
+      [onFocus],
+    )
+
+    const onInputChange = useCallback(
+      (event: React.ChangeEvent<HTMLInputElement>) => {
+        if (multiline) {
+          autoSize(event.target)
+        }
+
+        onChange?.(event)
+      },
+      [autoSize, multiline, onChange],
+    )
+
+    const onInputKeyDown = useCallback(
+      (event: React.KeyboardEvent<HTMLInputElement>) => {
+        if (event.code === 'Enter' || event.code === 'EnterNumpad') {
+          onEnterKeyDown?.(event)
+        }
+
+        onKeyDown?.(event)
+      },
+      [onEnterKeyDown, onKeyDown],
+    )
+
+    const internalInputProps = {
+      id,
+      value,
+      type,
+      name,
+      disabled,
+      onBlur: onInputBlur,
+      onChange: onInputChange,
+      onFocus: onInputFocus,
+      onKeyDown: onInputKeyDown,
+    }
+
+    const leadingIconsElements = leadingIcons.map((leadingIcon, index) => {
+      const elem = React.cloneElement(leadingIcon, { tabIndex: -1 })
+      return (
+        <LeadingIcon key={index} $index={index} $dense={dense}>
+          {elem}
+        </LeadingIcon>
+      )
+    })
+    const trailingIconsElements = trailingIcons
+      .slice() // Don't mutate the original array
+      .reverse()
+      .map((trailingIcon, index) => {
+        const elem = React.cloneElement(trailingIcon, { tabIndex: -1 })
+        return (
+          <TrailingIcon key={index} $index={index} $dense={dense} $multiline={multiline}>
+            {elem}
+          </TrailingIcon>
+        )
+      })
+
+    const clearAndFocusInput = useStableCallback(() => {
+      if (!inputRef.current) {
+        return
+      }
+
+      // React overrides the input value setter so we have to do shenanigans to set a value
+      // manually. See this answer for more info: https://stackoverflow.com/a/46012210/398302
+      const nativeInputValue = Object.getOwnPropertyDescriptor(
```

Comment:
I still think this is a quite odd thing to be doing at all. This input is controlled, i.e. its value should always be driven by props. You're doing things exactly backwards here. The more sensible approach is to deliver a change event to the normal handler (that is, just call `onChange`) saying that the value is now empty, and then focus the input, letting the parent change the props as needed.

Reply:
That's what I initially tried to do (just call the `onChange`), but then I realized the `event` parameter would be wrong, so I would have to change the signature of the `onChange` handler. Initially I tried making it so it accepts only a `value`, but then I saw that we're using some additional properties from the `change` event in forms (i.e., name), and then the only thing left was to make the parameter some weird thing. In the end I decided to just imperatively trigger the `change` event, since this functionality is internal to the TextField component itself.